waypoint_generator/src/waypoint_generator/waypoint_generator.cpp

```diff
@@ -0,0 +1,307 @@
+                // slow down in time
+                dist_accum += carma_wm::WorldModel::compute_euclidean_distance(
+                    centerline[j], 
+                    centerline[j + 1]);
+
+                if (dist_accum >= delta_d) {
+                    for (int k = j; k <= idx; k++) {
+                        // Linearly interpolate between speeds starting at that
+                        // point
+                        double dist = 
+                            carma_wm::WorldModel::compute_euclidean_distance(
+                                centerline[k], 
+                                centerline[idx + 1]);
+                        out[k] = 
+                            initial_v - 
+                            (1.0 - std::min(dist/dist_accum, 1.0)) *  
+                            delta_v;
+                    }
+                }
+
+            }
+        }
+
+        if (final_v > initial_v) {
+            // Speeding up
+            double delta_v = speeds[idx + 1] - speeds[idx];
+            double avg_v = delta_v/2.0;
+            double delta_t = delta_v/accel_limit;
+            double delta_d = avg_v * delta_t;
+
+            double dist_accum = 0;
+            for (int j = idx; j < speeds.size(); j++) {
+                // Iterate until we find a point that gives us enough delta_d to
+                // speed up acceptably
+                if (dist_accum >= delta_d) {
+                    for (int k = idx; k < j; k++) {
+                        // Linearly interpolate between speeds starting at that
+                        // point
+                        double dist = 
+                            carma_wm::WorldModel::compute_euclidean_distance(
+                                centerline[k], 
+                                centerline[j]);
+                        out[k] = 
+                            initial_v + 
+                            (1.0 - std::min(dist/dist_accum, 1.0)) *  
+                            delta_v;
+                    }
+                }
+
+                auto a = centerline[j];
+                auto b = centerline[j + 1];
+                dist_accum += carma_wm::WorldModel::compute_euclidean_distance(a, b);
+            }
+        }
+    }
+
+    return out;
+}
+
+std::vector<geometry_msgs::Quaternion> WaypointGenerator::compute_orientations(
+    const std::vector<lanelet::ConstLanelet> lanelets) const
+{
+    std::vector<geometry_msgs::Quaternion> out;
+
+    lanelet::BasicLineString2d centerline = carma_wm::WorldModel::concatenate_lanelets(lanelets);
+
+    std::vector<Eigen::Vector2d> tangents = carma_wm::WorldModel::compute_finite_differences(centerline);
+
+    Eigen::Vector2d x_axis = {1, 0};
+    for (int i = 0; i < tangents.size(); i++) {
+        geometry_msgs::Quaternion q;
+
+        // Derive angle by cos theta = (u . v)/(||u| * ||v||)
+        // TODO: std::acos is in radians, verify if this needs to be in degrees
```

Comment:
Does this TODO still apply?

Reply:
I actually probably ought to circle back on this one, it's still not entirely clear to me which is appropriate.